### PR TITLE
Transform encoder/decoder migration

### PR DESCRIPTION
**Pull request by sbesson**
This PR follows up on the RFE mentioned in https://github.com/openmicroscopy/omero-marshal/pull/12#issuecomment-243673798.

As `AffineTransformI` is now a proper OMERO object, it makes sense to maintain a dedicated set of encoder/decoder classes and update the shape encoder/decoder logic to delegate to these classes like we do elsewhere - see the ROI encoders/decoders for instance.

An additional complication with the transforms is the fact that these objects are introduced in OMERO 5.3+. In OMERO 5.1 and 5.2, OMERO transforms are simple string attributes. As we do not want encoder/decoder classes associated with the base `string` class, the proposal of  this PR is to:
- define transform encoder/decoder classes for the 2016-06 schema and onwards
- modify the top-level `__init__.py` logic to handle conditional imports by catching `ImportError` and skip modules when OMERO objects can not be imported.

The implications of catching `ImportError` should certainly be considered while reviewing this PR. The tests should remain green for all versions of OMERO.

### CODE REVIEW

**Comment by sbesson** [comment]
Discussed with @chris-allan who was not convinced by the `ImportError` strategy (as expected :) ). Given the challenges of this particular situation but also the needs to have a robust solution we can reuse in the future, the next step will be to use to an approach matching the handling of 5.1/5.2  Rectangle objects in `rectangle.py` by:
- defining an internal `omero_marshal.legacy.AffineTransformI`
- migrating all the string parsing/handling logic pre 5.3 to this class
- letting the the top-level `transform.py` import choose the relevant class (omero.model or legacy omero_marshal)

File: omero_marshal/decode/decoders/transform.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#
+# This software is distributed under the terms described by the LICENCE file
+# you can find at the root of the distribution bundle.
+# If the file is missing please request a copy by contacting
+# jason@glencoesoftware.com.
+#
+
+from ... import SCHEMA_VERSION
+from .annotation import Decoder
+try:
```
Comment (chris-allan) [inline omero_marshal/decode/decoders/transform.py]: Like we have done in the rectangle decoder, let's make explicit comments why we are doing this.

File: omero_marshal/encode/encoders/shape.py [hunk 2]
```diff
@@ -47,50 +50,15 @@ def set_linecap(self, v, obj):
...
-        return {
-            '@type': self.get_transform_type(),
-            'A00': a[0],
-            'A10': a[1],
-            'A01': a[2],
-            'A11': a[3],
-            'A02': a[4],
-            'A12': a[5],
-        }
+        from omero_marshal.legacy.AffineTransformI import AffineTransformI
```
Comment (chris-allan) [inline omero_marshal/encode/encoders/shape.py]: As above, let's explicitly comment why we are doing this.

File: omero_marshal/encode/encoders/transform.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#
+# This software is distributed under the terms described by the LICENCE file
+# you can find at the root of the distribution bundle.
+# If the file is missing please request a copy by contacting
+# jason@glencoesoftware.com.
+#
+
+from ... import SCHEMA_VERSION
+from .annotation import Encoder
+try:
```
Comment (chris-allan) [inline omero_marshal/encode/encoders/transform.py]: As above, let's explicitly comment why we are doing this.

**Comment by chris-allan** [comment]
Implementation looks great to me. Let's be a little more explicit in a few situations so we actually remember why we have these deviations and for which OMERO versions they apply and we should be good.

File: tests/unit/test_transform_decoders.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    ),
+    (
+        'rotate(60 50 100)',
+        'matrix(0.5 0.866025403784 -0.866025403784 0.5 '
+        '111.602540378 6.69872981078)'
+    ),
+]
+
+
+@pytest.mark.skipif(SCHEMA_VERSION != "2015-01", reason="legacy classes")
```
Comment (chris-allan) [inline tests/unit/test_transform_decoders.py]: Let's document in a little more detail why these are skipped and the scope of code that does or does not have coverage here.

**Comment by sbesson** [comment]
Last set of commits should include the full implementation of the proposal discussed in https://github.com/openmicroscopy/omero-marshal/pull/20#issuecomment-245268007 including inline documentation explaining some of the choices (`id = -1L`, details, conditional imports...)

File: omero_marshal/legacy/AffineTransformI.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+class AffineTransformI(object):
+    """
+    Class handling OMERO 5.1 and OMERO 5.2 transform string representations.
+    Its fields and methods are directly mirroring the
+    omero.model.AffineTransform class introduced in OMERO 5.3 to reduce the
+    complexity of the transform encoders/decoders
+    """
+
+    # Set the id to -1 to differentiate the marshalled transform from objects
```
Comment (chris-allan) [inline omero_marshal/legacy/AffineTransformI.py]: Only constants should be declared here. These are class variables as a result and unlikely what you had intended.

Comment (sbesson) [inline omero_marshal/legacy/AffineTransformI.py]: Good catch. Definitely an misuse of variables coming from a straight copy and paste of `omero_model_AffineTransformI.py`. Should be fixed by https://github.com/openmicroscopy/omero-marshal/pull/20/commits/cfdb7f253e7aa34224ec8a2a9d994c3ec4806b33

**Comment by chris-allan** [comment]
Cool. The only other feeling I had was that given we have standardised a little bit on the naming of the encoder and decoder packages we should `transform --> affinetransform` and `omero_marshal.legacy.AffineTransformI --> omero.marshal.legacy.affinetransform`.

Thoughts?

**Comment by sbesson** [comment]
@chris-allan: `transform -> affinetransform` in the encoders/decoders makes complete sense. Using `transform` was an oversight on my side, the only relevant place where it should be used is in the `Shape` encoder/decoders as it refers to the attribute name. I'll push a commit later to address this.

For the second one, I am in between multiple thoughts. Is there any risk of collision between root packages if we use `omero.marshal`? On the final namespace, I was thinking for symmetry reasons with the `omero.model` package there could be a case for `from omero.marshal.legacy import AffineTransformI` i.e. grouping all legacy classes into a top-level `__init__.py`?

**Comment by chris-allan** [comment]
If memory serves you'll get into some really weird stuff if you start trying to include an `omero` package in a separate module from `omero` itself. As `omero` is defined in OMERO.py and is not a namespace package or even an empty package, having a `omero` package in omero marshal is unlikely to work. You can certainly give it a try though.

**Comment by sbesson** [comment]
Agreed, I am definitely fine with not having a root `omero` package outside OMERO.py not only in terms of testing but also in terms of good practices. Re https://github.com/openmicroscopy/omero-marshal/pull/20#issuecomment-246405740, is the proposal `omero_marshal.legacy.affinetransform` rather than `omero.marshal.legacy.affinetransform` then?

**Comment by chris-allan** [comment]
Yes, apologies. Editing.

**Comment by sbesson** [comment]
Last commit should migrate the name of the legacy module as proposed in https://github.com/openmicroscopy/omero-marshal/pull/20#issuecomment-246442199 and refactor the unit tests to reduce their complexity, remove the conditional test skipping and isolate the testing of the legacy class.

A final remaining item I propose to tackle here with the corresponding unit tests is the migration/completion of the handling logic for the pre-OMERO 5.3 special cases transform strings (`none`, `invalid`) to the legacy class.

File: omero_marshal/legacy/affinetransform.py [hunk 6]
```diff
@@ -78,6 +89,9 @@ def __str__(self):
             unwrap(self._a12),
             ]))
 
+    def get_svg_transform(self):
```
Comment (chris-allan) [inline omero_marshal/legacy/affinetransform.py]: If we're going to store this and make it available in this manner maybe use a property, document it as public API and always make it available?

File: omero_marshal/legacy/affinetransform.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            unwrap(self._a00),
+            unwrap(self._a10),
+            unwrap(self._a01),
+            unwrap(self._a11),
+            unwrap(self._a02),
+            unwrap(self._a12),
+            ]))
+
+    @property
+    def get_svg_transform(self):
```
Comment (chris-allan) [inline omero_marshal/legacy/affinetransform.py]: We want the transform stored as a `_svg_transform` instance variable and this method to be an `svg_transform` property, no?

**Comment by sbesson** [comment]
Agreed on the roadmap. #18 is certainly a regression/critical bug that ought to be released as soon as possible in a patch release. This PR adds significant API in a backwards compatible manner thus incrementing the minor version number makes complete sense.

**Comment by chris-allan** [comment]
As 0.4.1 is out, merging.

**Pull request merged**